### PR TITLE
Typeahead

### PR DESCRIPTION
**Pull request by ian-leggett**
## Description
A new typeahead that we need in the referrals form, this will also eventually replace all the Vue typeaheads.

Trello - https://trello.com/c/xgH5rrE0/1095-build-react-typeahead-component

Note: We are only decorating a new dependancy (React Select) - https://react-select.com/home so I only have tests around the highlighting of matched characters as you type. We will test this in depth once ported to FE.

### CODE REVIEW

**Comment by ian-leggett** [comment]
> @ian-leggett For the async version, do you think that showing 'no advisors found' when a user first clicks in the type ahead could be misleading? Is it possible to show different text if they search and don't find something vs if they haven't searched anything yet?
> 
> Screenshot 2020-02-04 at 10 10 20

Good point but don't worry this is just an example of what you can use, you could remove this message if you like or pass any message you like.

File: src/typeahead/Typeahead.jsx [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    components: {
+      Option: (props) => {
+        return <Option {...props} inputValue={inputValue} />
+      },
+    },
+  }
+
+  const filterOptions = ({ label: labelToFilter }) =>
+    labelToFilter.toLowerCase().includes(inputValue) ||
+    (labelToFilter && labelToFilter.toLowerCase().includes(inputValue))
```
Comment (web-bert) [inline src/typeahead/Typeahead.jsx]: I think this could be clearer, I'm not sure what is trying to be accomplished  with the `labelToFilter &&` as it will have already blown up with the `labelToFilter.toLowerCase()` ? 

Maybe something like:
```JavaScript
function filterOptions({label}){
  const labelToFilter = label ? label.toLowerCase() : ''
  return labelToFilter.includes(inputValue)
}
```

File: src/typeahead/__stories__/Typeahead.stories.jsx [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  {
+    value: '379f390a-e083-4a2c-9cea-e3b9a08606a7',
+    label: 'Holly Collins',
+    subLabel: 'Heart of the South West LEP',
+  },
+  {
+    value: '379f390a-e083-4a2c-9cea-e3b9a08606a7',
+    label: 'Holly Collins',
+    subLabel: 'Heart of the South West LEP',
+  },
```
Comment (web-bert) [inline src/typeahead/__stories__/Typeahead.stories.jsx]: Is this intentional to have the same option twice?

File: src/typeahead/Typeahead.jsx (hunk 1 above)
Comment (ian-leggett) [reply src/typeahead/Typeahead.jsx]: Yeah I agree, I have refactored this and its much clearer to read now. See what you think 👍

File: src/typeahead/__stories__/Typeahead.stories.jsx (hunk 2 above)
Comment (ian-leggett) [reply src/typeahead/__stories__/Typeahead.stories.jsx]: It was at first as I was testing against duplicates but I have removed this now as we don't need it.

File: src/typeahead/Typeahead.jsx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    styles: customStyles,
+    components: {
+      Option: (props) => {
+        return <Option {...props} inputValue={inputValue} />
+      },
+    },
+  }
+
+  const filterOptions = ({ label: labelToFilter }) =>
+    labelToFilter ? labelToFilter.toLowerCase().includes(inputValue) : ''
```
Comment (web-bert) [inline src/typeahead/Typeahead.jsx]: should this return `false` instead of `''` as I assume it is going into an `array.filter` call?

Comment (ian-leggett) [reply src/typeahead/Typeahead.jsx]: Good spot this should have been false thanks 👍

File: src/typeahead/Highlighter.jsx [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import reactStringReplace from 'react-string-replace'
+import styled from 'styled-components'
+
+/* istanbul ignore file */
+const Highlighter = ({ optionLabel, searchStr }) => {
+  const StyledSpan = styled('span')`
+    font-weight: 900;
+  `
```
Comment (peterhudec) [inline src/typeahead/Highlighter.jsx]: This needs to go out of the `Highlighter` body, otherwise react will re-render it every time

File: src/typeahead/Typeahead.jsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+  const filterOptions = ({ label: labelToFilter }) =>
+    labelToFilter ? labelToFilter.toLowerCase().includes(inputValue) : false
+
+  return (
+    <>
+      <Label htmlFor={name}>
+        <StyledLabelText>{label}</StyledLabelText>
+        {hint && <StyledHintText>{hint}</StyledHintText>}
+      </Label>
```
Comment (peterhudec) [inline src/typeahead/Typeahead.jsx]: I would not bake the label into this component. The label and hint don't take part in the typeahead logic and don't need to be here. If, then I would create a separate `LabeledTypeahead`:
```js
const LabeledTypeahead = ({label, hint, ...props}) =>
  <Label htmlFor={props.name}>
    <LabelText>{label}</LabelText>
    {hint && <HintText>{hint}</HintText>}
    <Typeahead {...props} />
  </Label>
```

File: src/typeahead/__tests__/Typeahead.test.jsx [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      label="Standard options"
+      name="test_1"
+      options={transformOptions(selectOptions)}
+      placeholder="Search..."
+    />
+  )
+}
+
+describe('Typeahead', () => {
+  let wrapper
```
Comment (peterhudec) [inline src/typeahead/__tests__/Typeahead.test.jsx]: You don't need this `let`

File: src/typeahead/__tests__/Typeahead.test.jsx [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  describe('<Select>', () => {
+    describe('when searching with the characters "choc"', () => {
+      wrapper = mount(<WrappedSelectTypeahead />)
```
Comment (peterhudec) [inline src/typeahead/__tests__/Typeahead.test.jsx]: prefix this with `const` and then you won't need the `let` above

File: src/typeahead/__tests__/Typeahead.test.jsx [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      const highlightedTextArray = wrapper.find('[data-test-id="test-1"]')
+      const highlightedText = highlightedTextArray.map((x) => x.text())[0]
+      test('should highlight the matching string', () => {
+        expect(highlightedText).toEqual('Choc')
+      })
+    })
+  })
+})
```
Comment (rafenden) [inline src/typeahead/__tests__/Typeahead.test.jsx]: We can add a bit more tests, here's a bare minimum:

```suggestion
import React from 'react'
import { mount } from 'enzyme'
import transformOptions from '../transformOptions'

import Typeahead from '../Typeahead'
import Highlighter from '../Highlighter'

const OPTIONS = [
  { value: '1234', label: 'Chocolate', subLabel: 'mint' },
  { value: '5678', label: 'Strawberry', subLabel: 'Cream' },
  { value: '9876', label: 'Vanilla', subLabel: 'Vanilla' },
]

const BASIC_PROPS = {
  name: 'test',
  options: transformOptions(OPTIONS),
  menuIsOpen: true,
}

describe('Typeahead', () => {
  test('highlights typed text', () => {
    const selectWrapper = mount(
      <Typeahead {...BASIC_PROPS} inputValue="choc" />
    )

    expect(selectWrapper.find(Highlighter)).toHaveLength(3)

    expect(selectWrapper.text()).toEqual(
      'Select...Chocolate - mintStrawberry - CreamVanilla - Vanilla'
    )

    expect(
      selectWrapper
        .find(Highlighter)
        .at(0)
        .find('span')
        .text()
    ).toEqual('Choc')
  })
})

```

File: src/typeahead/transformOptions.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+/* istanbul ignore file */
+const transformOptions = (options) => {
+  return options.map(({ value, label, subLabel }) => {
+    return subLabel
+      ? {
+          value,
+          label: `${label} - ${subLabel}`,
```
Comment (rafenden) [inline src/typeahead/transformOptions.js]: Why not to simply use `label` only? What's the special use for `subLabel`?

File: src/typeahead/Typeahead.jsx (hunk 5 above)
Comment (ian-leggett) [reply src/typeahead/Typeahead.jsx]: So I would use this outside but it doesn't play nicely as if I wrap it with a label and hint etc the event handler gets eaten up inside the typeahead (bubbling issue). I see this similar to the "Field input text" component we have where you do get the label and hint. I can't think of a scenario when we would want this typeahead without a label.

File: src/typeahead/__tests__/Typeahead.test.jsx (hunk 6 above)
Comment (ian-leggett) [reply src/typeahead/__tests__/Typeahead.test.jsx]: Changed the tests :-)

File: src/typeahead/__tests__/Typeahead.test.jsx (hunk 8 above)
Comment (ian-leggett) [reply src/typeahead/__tests__/Typeahead.test.jsx]: Thanks for this, I have replaced your tests over mine but I can't think of how I can do more tests. Ideally we need it to simulate the typing of text in the input but it just won't play nicely. The above is much cleaner so I used it but it doesn't give me much more than what I had apart from asserting 3 options which we already know we have. It would be nice to see that when we search for "Choc" we have one option left and its characters "Choc" are highlighted. I would also like to test the async version but I can't type in the text field to fire the onChange handler :-(

File: src/typeahead/transformOptions.js (hunk 9 above)
Comment (ian-leggett) [reply src/typeahead/transformOptions.js]: So if you wanted to add the team beside the adviser such as `Rafal - DIT` you will need to concat the label and subLabel as we did in the Vue version of the typeahead. If there is no subLabel(team) we just render the label.

File: src/typeahead/__tests__/Typeahead.test.jsx (hunk 8 above)
Comment (rafenden) [reply src/typeahead/__tests__/Typeahead.test.jsx]: Looks like `react-select` doesn't test well https://stackoverflow.com/questions/41991077/testing-react-select-component

File: src/typeahead/transformOptions.js (hunk 9 above)
Comment (rafenden) [reply src/typeahead/transformOptions.js]: Well, a good reason to have them separately would be a consistent separator.

File: src/typeahead/Highlighter.jsx [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import reactStringReplace from 'react-string-replace'
+import styled from 'styled-components'
+import { FONT_WEIGHTS } from '@govuk-react/constants'
+
+/* istanbul ignore file */
```
Comment (rafenden) [inline src/typeahead/Highlighter.jsx]: Please remove.

File: src/typeahead/Typeahead.jsx [hunk 11]
```diff
@@ -0,0 +1,123 @@
+/* istanbul ignore file */
```
Comment (rafenden) [inline src/typeahead/Typeahead.jsx]: Please remove.

File: src/typeahead/__tests__/Typeahead.test.jsx [hunk 12]
```diff
@@ -0,0 +1,39 @@
+/* istanbul ignore file */
```
Comment (rafenden) [inline src/typeahead/__tests__/Typeahead.test.jsx]: No need for this in the test itself.

File: src/typeahead/transformOptions.js [hunk 13]
```diff
@@ -0,0 +1,16 @@
+/* istanbul ignore file */
```
Comment (rafenden) [inline src/typeahead/transformOptions.js]: Please remove.

File: src/typeahead/styles.js [hunk 14]
```diff
@@ -0,0 +1,82 @@
+/* istanbul ignore file */
```
Comment (rafenden) [inline src/typeahead/styles.js]: Please remove.

File: src/typeahead/__tests__/Typeahead.test.jsx (hunk 8 above)
Comment (rafenden) [reply src/typeahead/__tests__/Typeahead.test.jsx]: Although, you could add tests for the custom props, e.g. `reduceSize`

File: src/typeahead/Typeahead.jsx [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+  ...rest
+}) => {
+  const StyledLabelText = styled(LabelText)`
+    font-weight: ${reduceSize ? FONT_WEIGHTS.bold : FONT_WEIGHTS.regular};
+    ${MEDIA_QUERIES.TABLET} {
+      font-size: ${reduceSize ? FONT_SIZE.SIZE_16 : FONT_SIZE.SIZE_19};
+    },
+  `
+  const StyledHintText = styled(HintText)`
+    margin-bottom: ${SPACING.SCALE_2};
```
Comment (peterhudec) [inline src/typeahead/Typeahead.jsx]: These two need to go out of the component otherwise React would always re-render them

File: src/typeahead/Typeahead.jsx [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+const Typeahead = ({
+  name,
+  value,
+  label,
+  hint,
+  options,
+  inputValue,
+  reduceSize,
```
Comment (peterhudec) [inline src/typeahead/Typeahead.jsx]: I would remove the `reduceSize`. This is against the [Open–closed SOLID](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle) principle, which says:
> software entities (classes, modules, functions, etc.) should be open for extension, but closed for modification

You baked in a particular style variant instead of allowing it to be tweaked from the outside. If anyone would be in need of another variant, they would need to implement it here. I would do this exactly as the underlying library i.e. would allow to pass the whole `styles` with the styles yuo defined for our `Typeahead` as its default value. Then you could implement a `SmallTypeahead` by just wrapping `Typeahead` and overriding its `styles`.

File: src/typeahead/__tests__/Typeahead.test.jsx (hunk 8 above)
Comment (ian-leggett) [reply src/typeahead/__tests__/Typeahead.test.jsx]: After talking with Peter we have removed the label and hint out of this component as it made sense not to bake this in so these props are removed now.

File: src/typeahead/Typeahead.jsx (hunk 16 above)
Comment (ian-leggett) [reply src/typeahead/Typeahead.jsx]: Makes sense, I have now created a `<SmallTypeahead>` 👍

File: src/typeahead/Highlighter.jsx [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+const Highlighter = ({ optionLabel, searchStr }) => {
+  if (!optionLabel) {
+    return null
+  }
+  return reactStringReplace(optionLabel, searchStr, (matchedTxt, i) => (
+    <StyledSpan key={i} data-test-id={`test-${i}`}>
+      {matchedTxt}
+    </StyledSpan>
+  ))
+}
```
Comment (peterhudec) [inline src/typeahead/Highlighter.jsx]: This looks more complicated than it actually is. I would reverse the condition:
```js
const Highlighter = ({ optionLabel = null, searchStr }) =>
  optionLabel && reactStringReplace(optionLabel, searchStr, (matchedTxt, i) =>
    <StyledSpan key={i} data-test-id={`test-${i}`}>
      {matchedTxt}
    </StyledSpan>
  )
```

Comment (ian-leggett) [reply src/typeahead/Highlighter.jsx]: Good point, all done 👍

File: src/typeahead/Typeahead.jsx [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+    </components.Option>
+  )
+}
+
+const Typeahead = ({
+  name,
+  value,
+  options,
+  inputValue,
+  customStyles,
```
Comment (peterhudec) [inline src/typeahead/Typeahead.jsx]: I think this should be `styles` and not `customStyles` to preserve the `react-select` API.

File: src/typeahead/__tests__/Typeahead.test.jsx [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+]
+const BASIC_PROPS = {
+  name: 'test',
+  label: 'Advisers',
+  hint: 'Some hint',
+  options: transformOptions(OPTIONS),
+  menuIsOpen: true,
+}
+describe('Typeahead', () => {
+  let selectWrapper
```
Comment (peterhudec) [inline src/typeahead/__tests__/Typeahead.test.jsx]: You don't need to define a `let` here. Just define a `const wrapper = mount(...)` in each tests.

File: src/typeahead/__tests__/Typeahead.test.jsx [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+  test('renders options', () => {
+    selectWrapper = mount(<Typeahead {...BASIC_PROPS} inputValue="" />)
```
Comment (peterhudec) [inline src/typeahead/__tests__/Typeahead.test.jsx]: Prefix this with `const` and you wouldn't need the `let` above.

File: src/typeahead/__tests__/Typeahead.test.jsx [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+    expect(selectWrapper.find(components.Option)).toHaveLength(3)
+    expect(selectWrapper.find('MenuList').text()).toEqual(
+      'Chocolate - mintStrawberry - CreamVanilla - Vanilla'
+    )
+  })
+  test('highlights typed text', () => {
+    selectWrapper = mount(<Typeahead {...BASIC_PROPS} inputValue="choc" />)
```
Comment (peterhudec) [inline src/typeahead/__tests__/Typeahead.test.jsx]: Prefix this with `const` and you wouldn't need the `let` above.

File: src/typeahead/__tests__/Typeahead.test.jsx (hunk 7 above)
Comment (peterhudec) [reply src/typeahead/__tests__/Typeahead.test.jsx]: I can still see the `let` in there
